turbopack/crates/turbo-frozenmap/src/map.rs

```diff
@@ -0,0 +1,915 @@
+    where
+        K: Ord,
+    {
+        // Sort by key (stable sort preserves insertion order for equal keys)
+        entries.sort_by(|a, b| a.0.cmp(&b.0));
+
+        // Deduplicate, keeping the last value for each key.
+        // dedup_by removes the first argument when returning true, so we swap
+        // to keep the later (last) value in the earlier slot.
+        entries.dedup_by(|later, earlier| {
+            if later.0 == earlier.0 {
+                std::mem::swap(later, earlier);
+                true
+            } else {
+                false
+            }
+        });
+
+        Self::from_uniq_sorted_box(entries.into_boxed_slice())
+    }
+}
+
+impl<K: Ord, V> FromIterator<(K, V)> for FrozenMap<K, V> {
+    /// Creates a [`FrozenMap`] from an iterator of key-value pairs.
+    ///
+    /// If there are duplicate keys, the last value for each key is kept.
+    fn from_iter<T: IntoIterator<Item = (K, V)>>(iter: T) -> Self {
+        let mut entries: Vec<(K, V)> = iter.into_iter().collect();
+        if entries.is_empty() {
+            return Self::new();
+        }
+
+        // Sort by key (stable sort preserves insertion order for equal keys)
+        entries.sort_by(|a, b| a.0.cmp(&b.0));
+
+        // Deduplicate, keeping the last value for each key.
+        // dedup_by removes the first argument when returning true, so we swap
+        // to keep the later (last) value in the earlier slot.
+        entries.dedup_by(|later, earlier| {
+            if later.0 == earlier.0 {
+                std::mem::swap(later, earlier);
+                true
+            } else {
+                false
+            }
+        });
+
+        Self::from_uniq_sorted_box(entries.into_boxed_slice())
+    }
+}
+
+impl<K, V> From<BTreeMap<K, V>> for FrozenMap<K, V> {
+    /// Creates a [`FrozenMap`] from a [`BTreeMap`].
+    ///
+    /// This is more efficient than `From<HashMap<K, V>>` because [`BTreeMap`] already iterates in
+    /// sorted order, so no re-sorting is needed.
+    fn from(map: BTreeMap<K, V>) -> Self {
+        Self::from_uniq_sorted_iter(map)
+    }
+}
+
+impl<K, V, S> From<HashMap<K, V, S>> for FrozenMap<K, V>
+where
+    K: Ord,
+    S: BuildHasher,
+{
+    /// Creates a [`FrozenMap`] from a [`HashMap`].
+    ///
+    /// The entries are sorted by key during construction.
+    fn from(map: HashMap<K, V, S>) -> Self {
+        if map.is_empty() {
+            return Self::new();
+        }
+        Self::from_uniq_box_inner(map.into_iter().collect())
+    }
+}
+
+impl<K, V, S> From<IndexMap<K, V, S>> for FrozenMap<K, V>
+where
+    K: Ord,
+    S: BuildHasher,
+{
+    /// Creates a [`FrozenMap`] from an [`IndexMap`].
+    ///
+    /// The entries are sorted by key during construction.
+    fn from(map: IndexMap<K, V, S>) -> Self {
+        if map.is_empty() {
+            return Self::new();
+        }
+        Self::from_uniq_box_inner(map.into_iter().collect())
+    }
+}
+
```

Comment:
im not sure we should support these factories

the 'last one wins' behavior might be suprising and imho should be opt in.

Reply:
we should instead have `from_uniq_vec` and `from_uniq_sorted_vec`, `from_non_uniq_vec` (sorts and dedups)